### PR TITLE
Add qubit_subset option to BIPMapping pass

### PR DESCRIPTION
**Pull request by itoko**
### Summary
Add `qubit_subset` option to `BIPMapping` pass to provide an easier way to specify physical qubits to be used in mapping.

### Details and comments
Current BIP mapper (added at #6580) requires users to supply a reduced coupling map which contains only the physical qubits to be used (`qubit_subset`). That means it requires users to embed the circuit mapped on a reduced coupling map to the original coupling map as postprocessing. This seems a bit demanding. In addition, it also requires users to track virtual qubits -> qubits on reduced coupling map -> physical qubits and update `circuit._layout` by themselves to print the resulting circuit correctly.

This PR enables to do the above things in one stop (without requiring postprocessing). It can be done by supplying  `qubit_subset` with the original coupling map (reported by the backend). Users don't need to supply reduced coupling map any longer for specifying physical qubits to be used.

For example, suppose the following mapping,
```
circ = QuantumCircuit(3)
circ.cx(0, 1)
circ.cx(1, 2)
circ.cx(0, 2)
circ.measure_all()

qubit_to_use = [1, 3, 4]
coupling_map = CouplingMap([[0, 1], [1, 2], [1, 3], [3, 4]])
```
**After this PR**: The new `qubit_subset` option enables us to obtain the printable mapped circuit by one line:
```
transpiled = BIPMapping(coupling_map, qubit_subset=qubit_to_use)(circ)
print(transpiled)
```
```
ancilla_0 -> 0 ──────────────────────────────
                    ┌───┐         ░    ┌─┐   
      q_2 -> 1 ─────┤ X ├─X───────░────┤M├───
                    └─┬─┘ │       ░    └╥┘   
ancilla_1 -> 2 ───────┼───┼─────────────╫────
               ┌───┐  │   │ ┌───┐ ░     ║ ┌─┐
      q_1 -> 3 ┤ X ├──■───X─┤ X ├─░─────╫─┤M├
               └─┬─┘        └─┬─┘ ░ ┌─┐ ║ └╥┘
      q_0 -> 4 ──■────────────■───░─┤M├─╫──╫─
                                  ░ └╥┘ ║  ║ 
       meas: 3/══════════════════════╩══╩══╩═
                                     0  1  2 
```

**Before this PR**: The current interface requires dozens of lines to obtain the same mapped circuit:
```
reduced_coupling = coupling_map.reduce(qubit_to_use)  # == CouplingMap([[0, 1], [1, 2]])
transpiled = BIPMapping(reduced_coupling)(circ)

# fill layout with ancilla qubits
def fill_with_ancilla(layout):
    idle_physical_qubits = [q for q in range(coupling_map.size()) if q not in layout.get_physical_bits()]
    if idle_physical_qubits:
        qreg = QuantumRegister(len(idle_physical_qubits), name="ancilla")
        for idx, idle_q in enumerate(idle_physical_qubits):
            layout[idle_q] = qreg[idx]
        layout.add_register(qreg)
    return layout

# recover dag on original coupling map
layout = Layout({q: qubit_to_use[i] for i, q in enumerate(transpiled.qubits)})
for reg in transpiled.qregs:
    layout.add_register(reg)
property_set = {"layout": fill_with_ancilla(layout)}
recovered = ApplyLayout()(transpiled, property_set)

# recover layout
overall_layout = Layout({v: qubit_to_use[q] for v, q in transpiled._layout.get_virtual_bits().items()})
for reg in transpiled.qregs:
    overall_layout.add_register(reg)
recovered._layout = fill_with_ancilla(overall_layout)
print("recovered:", recovered)
```

### Alternatives considered
- Accept `PartialCouplingMap` instead of `CouplingMap`
`CouplingMap` class does not accept "holes" in physical qubits. So we may need to introduce another coupling class to allow holes in physical qubits, say `PartialCouplingMap`. However, `PartialCouplingMap` would be more likely to be a utility class for pass developers rather than interface class for pass users.

### CODE REVIEW

File: qiskit/transpiler/passes/routing/bip_mapping.py [hunk 1]
```diff
@@ -81,6 +80,7 @@ def __init__(
 
         Args:
             coupling_map (CouplingMap): Directed graph represented a coupling map.
+            qubit_subset (list[int]): Sublist of physical qubits to be used in the mapping.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/routing/bip_mapping.py:83]:
```suggestion
            qubit_subset (list[int]): Sublist of physical qubits to be used in the mapping.
                If None, all qubits in the coupling_map will be considered.
```

File: qiskit/transpiler/passes/routing/bip_mapping.py [hunk 2]
```diff
@@ -133,15 +135,16 @@ def run(self, dag):
...
+        if self.global_coupling_map is None:
             return dag
 
         if len(dag.qubits) > self.coupling_map.size():
             raise TranspilerError("More virtual qubits exist than physical qubits.")
 
         if len(dag.qubits) != self.coupling_map.size():
             raise TranspilerError(
-                "BIPMapping requires the number of virtual and physical qubits are the same."
+                "BIPMapping requires the number of virtual and physical qubits are the same. "
```
Comment (ajavadia) [inline qiskit/transpiler/passes/routing/bip_mapping.py:146]:
```suggestion
                "BIPMapping requires the number of virtual and physical qubits to be the same. "
```

File: qiskit/transpiler/passes/routing/bip_mapping.py [hunk 3]
```diff
@@ -105,18 +105,20 @@ def __init__(
...
         super().__init__()
-        self.coupling_map = copy.deepcopy(coupling_map)  # save a copy to modify
+        self.global_coupling_map = coupling_map
+        self.coupling_map = copy.deepcopy(coupling_map)  # copy to modify
+        self.global_qubit = qubit_subset
         if self.coupling_map is not None:
+            if self.global_qubit is None:
+                self.global_qubit = list(range(self.coupling_map.size()))
+            else:
+                self.coupling_map = self.coupling_map.reduce(self.global_qubit)
```
Comment (ajavadia) [inline qiskit/transpiler/passes/routing/bip_mapping.py:115]: what is the difference between `.reduce()` and `.subgraph()` methods here?

Comment (ajavadia) [reply qiskit/transpiler/passes/routing/bip_mapping.py:115]: #6738 fixes a bug on .subgraph() btw

**Review by ajavadia — COMMENTED** [review]
the approach looks good. And I agree with the mapping of qubits in the global_coupling_map and coupling_map to exist in BIPModel, so it can reason about physical error rates.

File: qiskit/transpiler/passes/routing/bip_mapping.py (hunk 3 above)
Comment (itoko) [reply qiskit/transpiler/passes/routing/bip_mapping.py:116]: `subgraph()` does not retain the order of passed qubits while `reduce()` does. For example (copied from https://github.com/Qiskit/qiskit-terra/issues/6736#issuecomment-881557566),
```
coupling = CouplingMap.from_line(6, bidirectional=False)
list(coupling.subgraph([4, 2, 3, 5]).get_edges())
>>> [(0, 1), (1, 2), (2, 3)]
list(coupling.reduce([4, 2, 3, 5]).get_edges())
>>> [(1, 2), (2, 0), (0, 3)]
```
We need the map from reduced qubits (in coupling_map) to global qubits (in global_coupling_map) to recover the global qubits from reduced qubits. Using `reduce()`, the map can be naturally represented as the list `qubit_subset` (and that's why the type of `qubit_subset` is `list` (not `set`) here).

**Review by ajavadia — APPROVED** [review]
looks good, and I'm happy that the pass doesn't call the 3 embedding stages now.

**Pull request merged**